app/master/cluster_master.py

```diff
@@ -213,6 +213,7 @@ def handle_request_for_new_build(self, build_params):
         if build_request.is_valid():
             build = Build(build_request)
             self._all_builds_by_id[build.build_id()] = build
+            build.generate_project_type()
```

Comment:
I mentioned this on Duke's PR as well, but having to call generate_project_type explicitly  opens the risk that project_type is never initialized and is None.  

I see we do some explicit None/ or isinstance checks in the code, but I am wondering if it would be better for consumers if the generate_project_type was called in the constructor.  

I also think separating the instantiation of the build object, and the call to generate_project_type makes the code look separated.  If we choose to keep the call to generate_project_type, can we put the call directly under the object creation?